File: WorldFeed/Server/src/WorldFeed.Common/Public/Iterators/IMultiLevelCursorIterator.cs
namespace WorldFeed.Common.Public.Iterators
{
    using System.Threading.Tasks;

    /// <summary>
    /// Twitter API misses some useful endpoints. Tweetinvi created multi level cursors that takes care of handling the complexity for you
    /// Twitter do not provide any endpoint to retrieve a collection of friends. It only supports retrieving friend ids
    /// </summary>
    public interface IMultiLevelCursorIterator<TParent, TItem> : IMultiLevelCursorIterator<TParent, TItem, string>
    {
    }

    /// <summary>
    /// Allow developers to iterate over multiple endpoints transparently.
    /// </summary>
    public interface IMultiLevelCursorIterator<TParent, TItem, TCursor>
    {
        bool Completed { get; }

        Task<IMultiLevelCursorIteratorPage<TParent, TItem, TCursor>> NextPageAsync();
    }
}

File: Sharebook/Client/src/app/core/Public/Iterators/IMultiLevelCursorIterator.ts
import {IMultiLevelCursorIteratorPage} from "./IMultiLevelCursorIteratorPage";

// Allow developers to iterate over multiple endpoints transparently.
export interface IMultiLevelCursorIterator<TParent, TItem, TCursor = any> {
  completed: boolean;

  nextPageAsync(): Promise<IMultiLevelCursorIteratorPage<TParent, TItem, TCursor>>;
}


// export interface IMultiLevelCursorIterator<TParent, TItem> extends IMultiLevelCursorIterator<TParent, TItem, string> {
// }

// export interface IMultiLevelCursorIterator<TParent, TItem, TCursor> {
//   Completed: boolean;
//
//   NextPageAsync(): Task<IMultiLevelCursorIteratorPage<TParent, TItem, TCursor>>
// }
